### DISCUSSION TITLE
FastLED multistrip setup gives "Error: invalid offset, value too big".

### DISCUSSION THREAD

**Discussion by miketanct** [earlephilhower/arduino-pico#1649] [Q&A]
With the following simple sketch, I get compile error in Arduino IDE (V2.1.1) when trying to set two or more pins with FastLED.addLeds, but no issues at all if set up with only one pin.

```
  #include "FastLED.h"
  
  #define NUM_LEDS_PER_STRIP 20
  CRGB leds1[NUM_LEDS_PER_STRIP];
  CRGB leds2[NUM_LEDS_PER_STRIP];

  void setup() {
    FastLED.addLeds<NEOPIXEL, 4>(leds1, NUM_LEDS_PER_STRIP);
    FastLED.addLeds<NEOPIXEL, 5>(leds2, NUM_LEDS_PER_STRIP); // an extra pin added here causes the compile error, but why?
  }
  
  void loop() {
  
  }

```
I've tried all the different ways for multi-strips with FastLED according to [this page](https://github.com/FastLED/FastLED/wiki/Multiple-Controller-Examples),  but all give the same error as long as more than one pin is used:

"{standard input}: Assembler messages:
{standard input}:873: Error: invalid offset, value too big (0x0000049C)" 

It's also not an issue when I select other boards not under this repository's list. 
Any help is appreciated.

**Comment by earlephilhower** [comment]
You could be running into an incompatibility w/the library and this RP2040 core, or a GCC compiler error.  Can you give more of a snapshot of the logs before the `Assembler error` to see what file is actually being compiled at the time?  If there's inline asm in the library it could be making assumptions about data layout that aren't correct (and hence some constant it's trying to load is too far for the manually written ARM Thumb instruction they used).

If it's a GCC bug, you could try changing the optimization level as a low-effort option.  Use the `Tools->Optimize` menu and try building at different `-O` levels.

At a worst case, there is the Arduino MBED RP2040 core you could try if all else fails.  They have a very different core and older GCC version last I checked which may not have the issue.

**Reply by miketanct** [reply]
Thank you for the response, Earle! Very honored to get such a fast reply from you.
I'm just an occasional tinkerer when it comes to programming and I'm not well-versed in programming terminology relating to compiling and assembly, so please pardon me for not understanding a large portion of your response. 😅

Nevertheless, here is the full log of the error -
```
In file included from e:\Documents\Arduino\libraries\FastLED\src/led_sysdefs.h:38,
                 from e:\Documents\Arduino\libraries\FastLED\src/FastLED.h:51,
                 from E:\Documents\Arduino\Pico_FastLED_multistrip_Error_Test\Pico_FastLED_multistrip_Error_Test.ino:2:
e:\Documents\Arduino\libraries\FastLED\src/platforms/arm/rp2040/led_sysdefs_arm_rp2040.h:87: warning: "cli" redefined
   87 | #define cli() (saved_interrupt_status = save_and_disable_interrupts())
      | 
In file included from C:\Users\miket\AppData\Local\Temp\arduino\sketches\031B191F947CA0CB9FEB770C728111CF\sketch\Pico_FastLED_multistrip_Error_Test.ino.cpp:1:
C:\Users\miket\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.3.2\cores\rp2040/Arduino.h:74: note: this is the location of the previous definition
   74 | #define cli() noInterrupts()
      | 
In file included from e:\Documents\Arduino\libraries\FastLED\src/led_sysdefs.h:38,
                 from e:\Documents\Arduino\libraries\FastLED\src/FastLED.h:51,
                 from E:\Documents\Arduino\Pico_FastLED_multistrip_Error_Test\Pico_FastLED_multistrip_Error_Test.ino:2:
e:\Documents\Arduino\libraries\FastLED\src/platforms/arm/rp2040/led_sysdefs_arm_rp2040.h:88: warning: "sei" redefined
   88 | #define sei() (restore_interrupts(saved_interrupt_status))
      | 
In file included from C:\Users\miket\AppData\Local\Temp\arduino\sketches\031B191F947CA0CB9FEB770C728111CF\sketch\Pico_FastLED_multistrip_Error_Test.ino.cpp:1:
C:\Users\miket\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.3.2\cores\rp2040/Arduino.h:73: note: this is the location of the previous definition
   73 | #define sei() interrupts()
      | 
In file included from e:\Documents\Arduino\libraries\FastLED\src/FastLED.h:75,
                 from E:\Documents\Arduino\Pico_FastLED_multistrip_Error_Test\Pico_FastLED_multistrip_Error_Test.ino:2:
e:\Documents\Arduino\libraries\FastLED\src/fastspi.h:172:23: note: '#pragma message: Forcing software SPI - no hardware SPI for you!'
  172 | #      pragma message "Forcing software SPI - no hardware SPI for you!"
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
{standard input}: Assembler messages:
{standard input}:873: Error: invalid offset, value too big (0x0000049C)

exit status 1

Compilation error: exit status 
```

When compiling with one of the FastLED.addLEDs line in my code removed, only the Assembler messages don't show up. All the log messages above it will still be present, but the compile still completes successfully.

I tried compiling with different optimization levels as per your suggestion but did not have any success using all of the levels available in Tools->Optimize.

Regarding the Arduino MBED RP2040 core, I have already compiled it successfully before trying your Arduino-pico core. The reason I am not choosing to use it is that I cannot figure out how to code it to utilize both cores on my Raspi Pico (I'm not even sure if it even supports dual core). Whereas, your excellent core library makes such easy work of utilizing both cores. It has provided me with the ideal solution for my project in which I had faced many difficulties initially resolving a multi-tasking requirement. So I am really grateful that you and your team have come up with this great repository. It makes things so much easier for a programming noob like me. Thank you very much!🙏

It's unfortunate that currently, this conflict with the FastLED library limits the ability for me to attach more than one LED strip to the microcontroller, but I can probably still work around it by rewiring the two led strips into a single strip and remapping in code to display them virtually as if they are two separate strips. Not a neat solution, but as the dual-core capability is more important in my project, I will still use yours instead of the MBED one. However, if you know how I can get two cores running with the MBED core, please point me to where I can find the instructions! 😁

Finally, I came across this other thread in the FastLED repository reporting the same error when compiling it with Arduino-pico - https://github.com/FastLED/FastLED/issues/1481. They are compiling thru CMake and since I'm basically an Arduino IDE user only for now, I don't understand most of the conversation there 😅 But perhaps there are cues there that you may find helpful to resolve this issue?

Thanks again Earle!

**Comment by FeuerSturm** [comment]
After just a quick look, only thing that is different is that you ommited the color order after the pin number, not sure if that causes problems.
I am using FastLED without problems and a lot of pins with this core.
Maybe try to add that or use NeoPXL8 lib instead.

**Reply by miketanct** [reply]
I did not specify the color order as the NEOPIXEL type does not allow changing the color order.
I also do not want to use the NEOPIXEL library as it hampers multitasking due to the way it blocks interrupts when running. Plus, FastLED has way more cooler effects that I can use! :) 

Anyhow, I tried your suggestion using WS2811 type in order to specify the color order, and discovered it has some impact on the error though it doesn't quite resolve the issue for me. What I found out was that I could compile with multiple strips successfully as long as each strip has a unique color order specified, but not when any two or more strips use the same color order. This means that using the following will compile with no issue: 
```
    FastLED.addLeds<WS2811, 2, GRB>(leds2, NUM_LEDS_PER_STRIP);
    FastLED.addLeds<WS2811, 3, GBR>(leds3, NUM_LEDS_PER_STRIP);
    FastLED.addLeds<WS2811, 4, BRG>(leds4, NUM_LEDS_PER_STRIP);
    FastLED.addLeds<WS2811, 5, BGR>(leds5, NUM_LEDS_PER_STRIP);
    FastLED.addLeds<WS2811, 6, RGB>(leds6, NUM_LEDS_PER_STRIP);
    FastLED.addLeds<WS2811, 7, RBG>(leds7, NUM_LEDS_PER_STRIP);
``` 
But using using as below will still result in the same error:
```
    FastLED.addLeds<WS2811, 2, GRB>(leds2, NUM_LEDS_PER_STRIP);
    FastLED.addLeds<WS2811, 3, GRB>(leds3, NUM_LEDS_PER_STRIP);
```

So I'm curious, what else could be different in your code that enables you to compile using multiple strips all with the same color order?
Thanks!

**Reply by FeuerSturm** [reply]
I am using WS2812B instead of NEOPIXEL or WS2811 and that works fine for me.
The NeoPXL8 lib is non blocking as well (unlike the bare Neopixel lib).

**Reply by earlephilhower** [reply]
I've just merged a toolchain update to GCC 12.3.  If you can run the `git master` version (be sure to re-run `./tools/get.py` to d/l the new toolchain) you might find something different happens since the compiled code will vary.

**Reply by haze1986** [reply]
Sorry, can you share how to update to the latest?

**Reply by earlephilhower** [reply]
https://github.com/earlephilhower/arduino-pico#installing-via-git

The official GCC 12.3 release will be sometime next week, when the IDE will pick it up automatically.

That said, no guarantees it changes anything.  But it's possible, especially if your code was hitting a GCC 10.x internal bug.

**Reply by horsepork** [reply]
@FeuerSturm Would you mind posting your code? I am able to get my code to compile so long as the led type/color order combo is unique.

@earlephilhower The new release did not fix the issue, fwiw.

**Comment by FeuerSturm** [comment]
@horsepork 

I have an online firmware generator setup,
just choose a channel for the pins and an led count and hit the large green button at the bottom, that will create a firmware with the settings.

This is for turning a Pi Pico into an ARGB LED Controller that works with SignalRGB's USB packets, so you can ignore most of the stuff and just focus on the FastLED settings in the generated INO:

[Firmware Generator](https://srgbmods.net/picoled/)

**Comment by glark-dave** [comment]
I'm facing this issue too; using two strips with the same RGB_ORDER for the Pico W causes:
```
{standard input}: Assembler messages:
{standard input}:840: Error: invalid offset, value too big (0x00000494)
```

E.g. this compiles fine:
```
  FastLED.addLeds<WS2812B, dataPin_1, RGB>(leds, 170);
  FastLED.addLeds<WS2812B, dataPin_2, GRB>(leds, 170);
```

but all my strips are GRB, and this:
```
  FastLED.addLeds<WS2812B, dataPin_1, GRB>(leds, 170);
  FastLED.addLeds<WS2812B, dataPin_2, GRB>(leds, 170);
```

results in the above compilation Assembler messages.

Any thoughts on how to move forward with this?

**Reply by Sreetej510** [reply]
see [https://github.com/Legion2/CorsairLightingProtocol/issues/346#issuecomment-1710347235](https://github.com/earlephilhower/arduino-pico/discussions/this), it helped with multiple led strips

**Reply by atlaiya** [reply]
me too, i just realised it happened when i added a third led strip with GRB when only the first two were there GRB and BRG it compiled and worked fine

**Reply by horsepork** [reply]
If you use a different led type, like ws2812, then it will still work even with the same color order. The led type you use in the addLeds function doesn't really seem to matter. This is a decent enough workaround until the underlying problem gets fixed

**Comment by Sreetej510** [comment]
see [https://github.com/Legion2/CorsairLightingProtocol/issues/346#issuecomment-1710347235](this), it helped with multiple led strips

**Reply by markbiasotti** [reply]
I too and running into this issue specifically with using FastLED's Fastled.add declaration for multiple strips and using the FastLED examples (ArrayOfLedArrays, MultiArrays, MultipleStripsInOneArray) Fastled works fine on a single strip but not when adding multiple strips. I've read this post and uninstalled it completely from IDE 1.8 and 2.2.1 (I first encountered using currently 2.2.1).  I've tried playing with the color order as has been suggested but to no avail. 

`#include <FastLED.h>

#define NUM_LEDS_PER_STRIP 36
CRGB redLeds[NUM_LEDS_PER_STRIP];
CRGB greenLeds[NUM_LEDS_PER_STRIP];
CRGB blueLeds[NUM_LEDS_PER_STRIP];

void setup() {

  FastLED.addLeds<WS2811, 16, BRG>(redLeds, NUM_LEDS_PER_STRIP);

  FastLED.addLeds<WS2811, 17, BRG>(greenLeds, NUM_LEDS_PER_STRIP);

  FastLED.addLeds<WS2811, 18, BRG>(blueLeds, NUM_LEDS_PER_STRIP);`

I have more description of my issue on the FastLED forum on Redit. [here](https://www.reddit.com/r/FastLED/comments/18bvvfp/fastled_support_for_adafruit_pr2040_scorpio/):

Here is the compile error:    

Error: invalid offset, value too big (0x00000490)


And this is the full compile output:

In file included from C:\Users\mbiasotti\Documents\Arduino\libraries\FastLED\src/led_sysdefs.h:38,
                 from C:\Users\mbiasotti\Documents\Arduino\libraries\FastLED\src/FastLED.h:51,
                 from C:\Users\MBIASO~1\AppData\Local\Temp\arduino_modified_sketch_142095\MultiArrays.ino:9:
C:\Users\mbiasotti\Documents\Arduino\libraries\FastLED\src/platforms/arm/rp2040/led_sysdefs_arm_rp2040.h:87: warning: "cli" redefined
   87 | #define cli() (saved_interrupt_status = save_and_disable_interrupts())
      | 
In file included from sketch\MultiArrays.ino.cpp:1:
C:\Users\mbiasotti\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.6.2\cores\rp2040/Arduino.h:74: note: this is the location of the previous definition
   74 | #define cli() noInterrupts()
      | 
C:\Users\mbiasotti\Documents\Arduino\libraries\FastLED\src/platforms/arm/rp2040/led_sysdefs_arm_rp2040.h:88: warning: "sei" redefined
   88 | #define sei() (restore_interrupts(saved_interrupt_status))
      | 
C:\Users\mbiasotti\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.6.2\cores\rp2040/Arduino.h:73: note: this is the location of the previous definition
   73 | #define sei() interrupts()
      | 
In file included from C:\Users\mbiasotti\Documents\Arduino\libraries\FastLED\src/FastLED.h:75:
C:\Users\mbiasotti\Documents\Arduino\libraries\FastLED\src/fastspi.h:172:23: note: '#pragma message: Forcing software SPI - no hardware SPI for you!'
  172 | #      pragma message "Forcing software SPI - no hardware SPI for you!"
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
{standard input}: Assembler messages:
{standard input}:1101: Error: invalid offset, value too big (0x00000490)
exit status 1
Error compiling for board Adafruit Feather RP2040 SCORPIO.

**Comment by markbiasotti** [comment]
UPDATE:  As previously mentioned, I have installed the Arduino MBED RP2040 core and it corrects this problem. I'm now able to run multiple strips with my Adafruit RP2040 Scorpio using FastLED. Thanks for the discussion as it has led me to a solution. Hope that Earle can get his core updated to correct the issue.  - Thanks

**Comment by markthomascampbell** [comment]
I'm weighing in with my experience too.  I too found this same problem when doing multiple LED arrays.  I am using a Raspberry Pi Pico, Arduino IDE 2.2.1, Raspberry Pi Pico/RP2040 3.6.3, FastLED 3.6.0.  In my project, I have 6 arrays, and tried multiple combinations of types.  I found that I could choose WS2811 and WS2812, but could not choose WS2812 and WS2812b.  The MBED one does work for me, but I don't know how to implement dual core on it either, as I do want to make use of the 2nd core for some processing.  Has there been any progress in understanding what the nature of the problem is?

**Reply by markthomascampbell** [reply]
I need to amend my previous statement about the MBED working.  It seems to work _Sometimes_, but I have a situation where the code compiles fine, and installs fine on my pico, and then runs fine.  Then I change only the brightness setting on the LEDs, and it compiles, installs, and then proceeds to lock up the Pico w/ the flasher of death on the activity light, and I need to reflash it in UF2 mode in order to get it to respond again.  I remember having an issue like this on another project where I had started using the MBED version, and it was advised that I switch to arduino-pico, and voila, it was working fine.  Unfortunately, for this project, that means that I'm stuck using either a version that sometimes works with specific brightness settings, or this version that only works if I can get past the error described in this thread.
@earlephilhower I'm kinda in a pickle.  Was wondering if you had made any head way on this issue?

**Comment by Andy2No** [comment]
@markthomascampbell How are you powering the Pi Pico and the LED strips?  Maybe it's drawing too much current when you change the brightness, reducing the voltage fed to the Pico.

**Reply by markthomascampbell** [reply]
@Andy2No There's not that many lights on my project, These are 17 total 5050 smd LEDs, and 48 3210 smd LEDs.  I am powering it straight from the Pico, which is powered from USB power via USB.  I can try supplementing the power (I was going to eventually, but this is a prototype), but I don't believe that's the problem, as I was able to power them at full brightness with a test script I used to test connectivity that just turned them all on, and that script doesn't lock up the Pico.  Additionally, in my main script, there's a start up sequence at the beginning of the script that only powers on 2 LEDs to start, and with the MBED Pico & higher brightness setting, the Pico locks up before it even reaches that point.  All I have to do to fix it, is turn down the brightness, and the Pico works as intended, with all of the lights lighting up with no issues.  The Arduino-Pico doesn't have any such issues, except the issue that this thread addresses.  I can partly get around it by doing most of my arrays as different types, but I have only found 3 types that work as a workaround, and I need 6.

**Reply by markthomascampbell** [reply]
@Andy2No I added in a separate 5V power supply to the board, and confirmed that the problem still remains.

**Comment by ricardok-sp** [comment]
I have the same messages using a Pico, Arduino IDE 2.30, Fastled 3.6.0 with two WS2812b strips. Sometimes I get the same message when using only one strip too:

`{standard input}: Assembler messages:
{standard input}:6101: Error: invalid offset, value too big (0x00000530)
{standard input}:6128: Error: invalid offset, value too big (0x000004EC)
`

**Edit:** the workaround of setting the second strip as RGB, instead of GRB (as the first), does work. 
**Edit2:** I really don't know what happens: didn't change the code, and now the error happens with either 1 or 2 strips...

**Comment by Gavin-Perry** [comment]
One quick work around for 2 strips is to use separate code for the second strip in the other core processor since RP2040 has 2 cores. I've used the second core in a motor control program that needed to run 4 motors and could only handle 2 in the code.

add void setup1() { // 2nd processor setup
  // Your code here
}
// and
void loop1() {  // 2nd processor code
  // your code for second strip here
}

Doesn't solve the sei and cli issue but at least now it's just a warning

**Reply by miketanct** [reply]
Thanks for the suggestion. It is a good workaround. Unfortunately, I cannot apply it to my project because one of the cores is already dedicated to checking and processing RF signals from a receiving RF module. So I'm only left with the other core to handle the FastLED processes. 

Anyway, I managed to get my project working by splitting one LED strip virtually into two groups using code even though I declared only one LED strip. See my project details at - https://www.hackster.io/miketanct/wireless-scoreboard-display-for-drone-soccer-or-other-games-f10ade

**Comment by KStandiford** [comment]
Wow. It's still a problem, and the MBED core works.

**Comment by MaEtUgR** [comment]
I don't fully understand this problem but experimented a bit to find a workable solution for the current needs of our project:
- PlatformIO a few days ago switched to FastLED 3.7.0 which gets loaded even if you previously used 3.6.0 because dependencies are by default added as minimum version requirement. I switched to using `fastled/FastLED@3.6.0` without the `^` before the version number to force version 3.6.0 which at least compiles and successfully runs again the previously working code.
https://github.com/atopile/swoop/pull/24
- Moving the LED initialization into a separate constructor leads to the same compile error and I had to work around that by:
  - Using a setup function in the class called from the Arduino's global `setup()` function instead of the constructor, no idea why. Initialization order?
  - I had to make a class member LED controller `NEOPIXEL<LED_DATA_OUT_PIN> _led_controller;` and call `FastLED.addLeds(&_led_controller, _leds, NUM_LEDS);` in the setup function instead of the previous call `FastLED.addLeds<NEOPIXEL, LED_DATA_OUT_PIN>(leds, NUM_LEDS);` which was in global scope.
https://github.com/atopile/swoop/pull/25

Omitting even just one point above leads to the compile error 🤯

**Comment by biomurph** [comment]
This is still an issue in the Arduino 2.3.2 I get this error


   172 | #      pragma message "Forcing software SPI - no hardware SPI for you!"
        |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  {standard input}: Assembler messages:
  {standard input}:3289: Error: invalid offset, value too big (0x00000530)
  {standard input}:3316: Error: invalid offset, value too big (0x000004EC)

But then, when I use the mbed core, it does compile!

**Comment by darrenbeckley** [comment]
I was also getting the "Error: invalid offset, value too big" error even when building for a single strip. I found setting FASTLED_RP2040_CLOCKLESS_M0_FALLBACK to zero before including FastLED.h solved the problem for me. See [my comment](https://github.com/FastLED/FastLED/issues/1629#issuecomment-2275336557) for more details.

Maybe this will resolve the issue with multiple strips as well?

**Reply by JonnyHaystack** [reply]
If enough DMA channels are available then yeah it should. Ideally on RP2040 you would only use the DMA implementation, as it is superior.

**Reply by pdcook** [reply]
This is exactly what I needed to fix my issue. Thanks so much!
